app/scripts/controllers/network/provider-api-tests/createClient.test.js

```diff
@@ -0,0 +1,216 @@
+/**
```

Comment:
not the biggest fan, because this makes it IMO more difficult to tell what the actual differences are. With this PR, you can search for providerType === 'infura' and find every case where its different. If they are seperate files, over time they will continue to drift and the differences become less clear. I would like us to move towards unifying these 2 interfaces and implementations, and starting with the tests is IMO a good way to get us there. 

One thing that I could get behind is to move all the things that work differently into 2 seperate test files (`createInfuraClient.test.js` and `createJsonRpcClient.test.js`), and leave the things that work the same in a `createClient.test.js`. I'm not sure the juice is worth the squeeze on that one though.

Reply:
> not the biggest fan, because this makes it IMO more difficult to tell what the actual differences are. With this PR, you can search for providerType === 'infura' and find every case where its different.
>
> ...
>
> I would like us to move towards unifying these 2 interfaces and implementations, and starting with the tests is IMO a good way to get us there.

Gotcha. I guess the only other argument I would make here about keeping all the tests in one file vs. two is that it creates more code complexity and that may end up impacting maintainability in the future. I've found, when writing tests, that every time I use loops and logic and other mechanisms that I would usually use in an implementation file to make code more compact, it pretty much always ends up biting me when I go to update those tests later. If you want a straightforward list of what this function does, now you have to go through each test and apply the logic yourself, and it can be a burden to do that.

That said, I see your point about being able to tell the differences more easily if you have the logic right in front of you instead of having to do the comparison between two files yourself. Obviously right now we don't have that much logic in these tests, so it's not a big deal. Also, it sounds like you have a plan for unifying all of this and maybe this will end up getting simplified or something. So, maybe this is just something we watch out for in the future.

Reply:
For sure I share in your frustrations of having to deal with overly abstract testing setups, this is also a worry of mine. Theres a delicate tradeoff between additional complexity Vs. how much code needs to be updated for any given change. More places to update leads to greater likelyhood of mistakes being made and not caught. More complexity makes it take longer before someone is able to update / fix any of that code. The list of tradeoffs goes on. Context is important in determining how much of each is appropriate, and in the context of testing, I dont think we should aim to minimize complexity at the cost of extreme duplication, just as I dont think we should aim to optimize for lowering lines of code. We should aim to group functionality into tests that if they break, they are broken in relation to the grouped functionality.

That being said, I think due to the nature of these tests (broad coverage of general functionalities), I think this level of complexity and abstractness is acceptable, and in this case, I think it actually makes them easier to maintain, since changes to the client / middleware require far fewer updates to the tests. I'm certainly not advocating for over-valueing reduction in lines of code, but I believe this to be more of a grouping tests based around some general functionalities. Unless you are changing those general functionalities, you should never end up with one of these tests breaking on you. If you are adding a new method to this set but it doesnt pass, putting it in a different test file makes it clear that there is a deviation from the standard client middleware functionality. 

As for unifying it all, I think we can do this at the createXYZClient interface level to start with at the same time as we migrate this code to controllers repo. This will overall minimize the change set introduced when later integrating controllers/network-controller with extension.

Reply:
Hmm. Alternatively, we could take the loop contents here and move them to a "shared client tests" module, and use that in both `createJsonRpcClient.test.js` and `createinfuraClient.test.js`.

It would be nice to keep the convention of the test module matching the name fo the module under test. But in addition to that, I had planned on adding some tests that would only be applicable to the Infura client (for the Infura client specific middleware). That would better fit in a dedicated module for that client as well.

Reply:
> As for unifying it all, I think we can do this at the createXYZClient interface level to start with at the same time as we migrate this code to controllers repo. This will overall minimize the change set introduced when later integrating controllers/network-controller with extension.

Even if we want to unify these interfaces, I am going to push back slightly on this and argue that we should hold off on doing that. Not only would duplicating the tests / test helpers make this PR more focused, as well as the PR that will end up doing the unification, but it also gives us room to change our mind in case the separation does turn out to be useful (as Mark points out).

Reply:
@BelfordZ I pushed a commit which restored `createInfuraClient.test.js` and `createJsonRpcClient.test.js`. The bulk of the tests in `shared-tests.js` as you had them; the individual files just call a function. I did pull out a couple of tests, however, that seemed to be Infura-specific.